dvc/stage_cache.py

```diff
@@ -0,0 +1,72 @@
+import os
+import json
+import hashlib
+
+from dvc.utils.fs import makedirs
+
+
+def _sha256(string):
+    return hashlib.sha256(string.encode()).hexdigest()
+
+
+def _get_hash(stage):
+    if not stage.cmd or not stage.deps or not stage.outs:
+        return None
+
+    string = _sha256(stage.cmd)
+    for dep in stage.deps:
+        if not dep.def_path or not dep.get_checksum():
+            return None
+
+        string += _sha256(dep.def_path)
+        string += _sha256(dep.get_checksum())
+
+    for out in stage.outs:
+        if not out.def_path:
+            return None
+
+        string += _sha256(out.def_path)
+
+    return _sha256(string)
+
+
+def _get_cache(stage):
+    return {
+        "cmd": stage.cmd,
+        "deps": {dep.def_path: dep.get_checksum() for dep in stage.deps},
+        "outs": {out.def_path: out.get_checksum() for out in stage.outs},
+    }
+
+
+def _get_cache_path(stage):
+    sha = _get_hash(stage)
+    if not sha:
+        return None
+
+    cache_dir = os.path.join(stage.repo.cache.local.cache_dir, "stages",)
+
+    return os.path.join(cache_dir, sha[:2], sha)
```

Comment:
if sha is 123456, then path would be `12/123456` and not `12/3456` as it is in our regular data cache. the reason is do it more like git does, though I don't see any practical advantage right now. Might revert to `12/3456`, just to be consistent.

Reply:
+1 for being consistent with our cache.